src/gui/notifications/dbusnotifier.cpp

```diff
@@ -0,0 +1,230 @@
+        return QUrl::fromLocalFile(h->rootPath());
+    }
+}
+
+void DBusNotifier::notificationClosed(uint id, uint /*reason*/)
+{
+    activeNotifications.remove(id);
+}
+
+void DBusNotifier::actionInvoked(uint id, const QString &action_key)
+{
+    if (!activeNotifications.contains(id)) {
+        return;
+    }
+    const Context& context = activeNotifications[id];
+    if (action_key == ACTION_NAME_DEFAULT) {
+        if (context.widget) {
+            context.widget->activateWindow();
+        }
+    }
+    else if (action_key == ACTION_NAME_OPEN_FINISHED_TORRENT) {
+        const BitTorrent::TorrentHandle* h = context.torrent;
+        // if there is only a single file in the torrent, we open it
+        if (h) {
+            QDesktopServices::openUrl(getUrlForTorrentOpen(h));
+        }
+    }
+}
+
+void DBusNotifier::showNotification(NotificationKind kind, const QString &title, const QString &message, const Context *context)
+{
+    QStringList actions;
+//    actions.append("default"); // This does not work with Plasma, it shows a button
+//    actions.append(ACTION_NAME_DEFAULT);
+    QVariantMap hints;
+    bool storeNotification = false;
+    int timeout = 0;
+    switch (kind)
+    {
+    case Notifier::TorrentFinished:
+        if (!capabilities.testFlag(BodyHyperlinks)) { // we need a button
+            actions.append(ACTION_NAME_OPEN_FINISHED_TORRENT);
+            actions.append(tr("Open", "Open donwloaded torrent"));
+            storeNotification = true;
+        }
+        hints[QLatin1String("category")] = QLatin1String("transfer.complete");
+        if (capabilities.testFlag(ActionIcons)) {
+            hints[QLatin1String("action-icons")] = true;
+        }
+        break;
+    case Notifier::TorrentIOError:
+        hints[QLatin1String("category")] = QLatin1String("transfer.error");
+        hints[QLatin1String("urgency")] = 2;
+        hints[QLatin1String("resident")] = true;
+        break;
+    case Notifier::UrlDownloadError:
+        hints[QLatin1String("category")] = QLatin1String("network.error");
+        hints[QLatin1String("urgency")] = 2;
+        hints[QLatin1String("resident")] = true;
+        break;
+    case Notifier::SearchFinished:
+         hints[QLatin1String("urgency")] = 0;
+         timeout = -1;
+         storeNotification = true;
+         break;
+    default:
+        break;
+    }
+    const QString APPLICATION_NAME = QLatin1String("qBittorrent");
+    hints[QLatin1String("desktop-entry")] = APPLICATION_NAME;
+
+    QDBusPendingReply<uint> reply = notificationsService->Notify(APPLICATION_NAME, 0, QLatin1String("qbittorrent"), // icon name
+                    title, message, actions, hints, timeout);
+    reply.waitForFinished();
+    if (storeNotification && reply.value() != 0) {
+        Context ctx = context ? *context : Context();
+        if (!ctx.widget) {
+            ctx.widget = mainWindow;
+        }
+        activeNotifications.insert(reply.value(), ctx);
```

Comment:
Thank you for the explanation! And I'm sorry that you night sleep was
disturbed by me. I think it is time for me to point out that my intention
is to make qBittorent better, as well as your.

I already read the code, of course. You already pointed out the big
difference between the handle in libtorrent and the TorrentHandle class:
libtorrent signals by throwing an exception if an operation was not
completed, while TorrentHandle does not. Beginning with SAFE_CALL, which
eats exceptions and ending with objects which can be deleted.

I do not argue that the solution to the libtorrent handles (Session +
torrent objects with Qt-ised interface) is bad, no! I'm arguing that there
may be a better solution for the ownership problem. For instance, I would
prefer non-copyable handle classes, given to clients by reference.

On 25 October 2015 at 07:44, Vladimir Golovnev <notifications@github.com>
wrote:

> In src/gui/notifications/dbusnotifier.cpp
> <https://github.com/qbittorrent/qBittorrent/pull/3810#discussion_r42941784>
> :
>
> > +         break;
> > +    default:
> > +        break;
> > +    }
> > +    const QString APPLICATION_NAME = QLatin1String("qBittorrent");
> > +    hints[QLatin1String("desktop-entry")] = APPLICATION_NAME;
> > +
> > +    QDBusPendingReply<uint> reply = notificationsService->Notify(APPLICATION_NAME, 0, QLatin1String("qbittorrent"), // icon name
> > +                    title, message, actions, hints, timeout);
> > +    reply.waitForFinished();
> > +    if (storeNotification && reply.value() != 0) {
> > +        Context ctx = context ? *context : Context();
> > +        if (!ctx.widget) {
> > +            ctx.widget = mainWindow;
> > +        }
> > +        activeNotifications.insert(reply.value(), ctx);
>
> That is why TorrentHandle is not even a classical handle, not a classical
> smart pointer. What is it? Can't it become better by gaining the proper
> ownership semantic?
>
> Do you really think that only "classical" handle or "classical" smart
> pointer are the right patterns?
> Okay, I am forced to spend my time and explain everything from the
> beginning...
>
> The handle for me, I guess, has common meaning: it is something that allow
> me to access an object while I own the handle.
>
> That is torrent handle in libtorrent.
>
> And the idea that handle may become invalid (and crash the program,
> actually) without my actions is kind of surprising, don't you agree?
>
> I don't agree with this.
> This is one reason why I decided to abandon this architecture and to hide
> all interaction with libtorrent::torrent_handle inside our TorrentHandle.
> libtorrent::torrent_handle may become invalid at any time even directly
> between calling is_valid() (returning true) and next method.
> The second reason we need to store any additional data in TorrentHandle
> class.
>
> If we consider a remote resource, then any operation on its handle may
> fail, naturally. But this has to be stated in the operation semantic (say,
> throw a specified exception). However, there is no any error reporting in
> the TorrentHandle semantic.
>
> Any libtorrent::torrent_handle method can throw an exception. If we
> promote a similar architecture to the upper levels of the application, the
> qBittorrent code will grow at the expense of having to use try-catch blocks
> everywhere. In this case, these blocks will not incur any benefit, in
> addition to suppression of exceptions (if you look, in which case
> torrent_handle may become invalid).
>
> So in current qBittorrent...
> Session creates TorrentHandle instances (one for each torrent) and *owns*
> them. The others can only obtain a pointer to one of these objects. No need
> of smart pointers here.
> The state of TorrentHandle objects is updated "discretely", i.e. we assume
> that for some period of time the object remains unchanged. This greatly
> simplifies the interaction with the application BitTorrent engine.
> More info can try to understand it yourself (read the code or ask some
> specific questions).
>
> If you're against using the name TorrentHandle, you can create a relevant
> Issue on this bugtracker.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/qbittorrent/qBittorrent/pull/3810/files#r42941784>.
>